docs/RFCS/schema_gossip.md

```diff
@@ -0,0 +1,34 @@
+- Feature Name: schema_distribution
+- Start Date: 2015-07-20
+- RFC PR:
+- Cockroach Issue:
+
+# Summary
+
+This RFC suggests implementing eventually-consistent replication of the SQL schema to all Cockroach nodes in a cluster using gossip.
+
+# Motivation
+
+In order to support performant SQL queries, each gateway node must be able to address the data requested by the query. Today this requires the node to read `TableDescriptor`s from the KV map, which we believe (though we haven't measured) will cause a substantial performance hit which we can eliminate by actively gossiping the necessary metadata.
+
+# Detailed design
+
+Distribution of the schema metadata will be modeled after the current configuration distribution mechanism. Whenever a new `TableDescriptor` is written to the KV map, the node holding the leader lease for that key's range will add the `TableDescriptor` to gossip under a key derived from the `TableDescriptor`'s `Name` with a TTL TBD.
+
+On the read side, `sqlserver.Server.getTableDesc`'s behaviour will change such that it reads from gossip rather than the KV store. This will require plumbing a referece to the `Gossip` instance down to the `sqlserver.Server` instance.
+
+# Drawbacks
+
+This is slightly more complicated than the current implementation. Because the schema is eventually-consistent, how do we know when migrations are done? We'll have to count on the TTL, which feels a little dirty.
+
+# Alternatives
+
+We could stick to the current implementation, perhaps using inconsistent reads for better performance. This will be strictly less performant than the gossip approach but will be more optimal in that nodes will only need to carry in memory schema information that they themselves need.
+
+We could have a special metadata range which all nodes have a replica of. This would probably result in unacceptable read and write times which induce lots of network traffic.
+
+We could implement this on top of non-voting replicas (which we don't yet support). That would give us eventual consistency without having to outside of raft, but enforcement of schema freshness remains an open question.
```

Comment:
@BramGruneir I don't believe this solution would work.

Things I'd be worried about:

- There's currently no exposed mechanism for looking up the involved ranges to do necessary PUT ops. We'd need to scan the meta records ourselves in the txn, which would be an unwelcome abstraction violation.
- Lots of schema duplication if (as you'd expect) many ranges on a single node would be part of the same table. The write amplification through the system in order to do a schema change would be dramatic.

What I think won't work:

The real problem isn't getting the schema to the ranges which make up a table. Where we need the schema is at the gateway node which is processing the SQL. Which means we need the schemas at every gateway node. But there's actually no way in Cockroach to know the full set of nodes, or for that matter, to effectively wait until all are contacted (it would make schema changes impossible in the event any node were down).